File: examples/typescript/destructive-guard.ts
#!/usr/bin/env -S npx tsx
/**
 * destructive-guard.ts — Claude Code PreToolUse hook in TypeScript
 *
 * Blocks rm -rf /, git reset --hard, git clean -fd, and similar
 * destructive commands. Exit code 2 = block, 0 = allow.
 *
 * Run with: npx tsx destructive-guard.ts
 * Or compile: npx tsc destructive-guard.ts && node destructive-guard.js
 */

interface HookInput {
  tool_input: {
    command?: string;
  };
}

const DANGEROUS_PATTERNS: RegExp[] = [
  /\brm\s+.*-rf\s+(\/|~\/?\s*$|\.\.\/)/,
  /\bgit\s+reset\s+--hard/,
  /\bgit\s+clean\s+-[a-zA-Z]*f/,
  /\bgit\s+checkout\s+--force/,
  /\bchmod\s+(-R\s+)?777\s+\//,
  /\bfind\s+\/\s+-delete/,
  /Remove-Item.*-Recurse.*-Force/,
  /--no-preserve-root/,
  /\bsudo\s+mkfs\b/,
];

async function main(): Promise<void> {
  let data = '';
  for await (const chunk of process.stdin) {
    data += chunk;
  }

  let input: HookInput;
  try {
    input = JSON.parse(data);
  } catch {
    process.exit(0); // Don't block on parse error
  }

  const cmd = input.tool_input?.command;
  if (!cmd) {
    process.exit(0);
  }

  // Skip echo/printf context
  const trimmed = cmd.trimStart().toLowerCase();
  if (trimmed.startsWith('echo ') || trimmed.startsWith('printf ')) {
    process.exit(0);
  }

  for (const pattern of DANGEROUS_PATTERNS) {
    if (pattern.test(cmd)) {
      process.stderr.write(`BLOCKED: Dangerous command detected\nCommand: ${cmd}\n`);
      process.exit(2);
    }
  }

  process.exit(0);
}

main();

File: examples/go/destructive_guard.go
// destructive_guard.go — Claude Code PreToolUse hook in Go
//
// Blocks rm -rf /, git reset --hard, git clean -fd, and similar
// destructive commands. Exit code 2 = block, 0 = allow.
//
// Build: go build -o destructive-guard destructive_guard.go
// Usage in settings.json:
//   {"type": "command", "command": "/path/to/destructive-guard"}
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"regexp"
	"strings"
)

type HookInput struct {
	ToolInput struct {
		Command string `json:"command"`
	} `json:"tool_input"`
}

var dangerousPatterns = []*regexp.Regexp{
	regexp.MustCompile(`\brm\s+.*-rf\s+(/|~/?\s*$|\.\./)`),
	regexp.MustCompile(`\bgit\s+reset\s+--hard`),
	regexp.MustCompile(`\bgit\s+clean\s+-[a-zA-Z]*f`),
	regexp.MustCompile(`\bgit\s+checkout\s+--force`),
	regexp.MustCompile(`\bchmod\s+(-R\s+)?777\s+/`),
	regexp.MustCompile(`\bfind\s+/\s+-delete`),
	regexp.MustCompile(`Remove-Item.*-Recurse.*-Force`),
	regexp.MustCompile(`--no-preserve-root`),
	regexp.MustCompile(`\bsudo\s+mkfs\b`),
}

func main() {
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		os.Exit(0) // Don't block on read error
	}

	var input HookInput
	if err := json.Unmarshal(data, &input); err != nil {
		os.Exit(0) // Don't block on parse error
	}

	cmd := input.ToolInput.Command
	if cmd == "" {
		os.Exit(0)
	}

	// Skip if command is in an echo/printf context
	lower := strings.ToLower(cmd)
	if strings.HasPrefix(strings.TrimSpace(lower), "echo ") ||
		strings.HasPrefix(strings.TrimSpace(lower), "printf ") {
		os.Exit(0)
	}

	for _, pattern := range dangerousPatterns {
		if pattern.MatchString(cmd) {
			fmt.Fprintf(os.Stderr, "BLOCKED: Dangerous command detected\nCommand: %s\n", cmd)
			os.Exit(2)
		}
	}

	os.Exit(0)
}
